azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_sonarcloud.go

```diff
@@ -0,0 +1,52 @@
+package serviceendpoint
+
+import (
+	"github.com/google/uuid"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/microsoft/azure-devops-go-api/azuredevops/v6/serviceendpoint"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils/converter"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils/tfhelper"
+)
+
+// ResourceServiceEndpointSonarCloud schema and implementation for SonarCloud service endpoint resource
+func ResourceServiceEndpointSonarCloud() *schema.Resource {
+	r := genBaseServiceEndpointResource(flattenServiceEndpointSonarCloud, expandServiceEndpointSonarCloud)
+
+	r.Schema["token"] = &schema.Schema{
+		Type:             schema.TypeString,
+		Required:         true,
+		Sensitive:        true,
+		DiffSuppressFunc: tfhelper.DiffFuncSuppressSecretChanged,
+		ValidateFunc:     validation.StringIsNotWhiteSpace,
+		Description:      "Authentication Token generated through SonarCloud (go to My Account > Security > Generate Tokens)",
+	}
+	// Add a spot in the schema to store the token secretly
+	stSecretHashKey, stSecretHashSchema := tfhelper.GenerateSecreteMemoSchema("token")
+	r.Schema[stSecretHashKey] = stSecretHashSchema
+
+	return r
+}
+
+// Convert internal Terraform data structure to an AzDO data structure
+func expandServiceEndpointSonarCloud(d *schema.ResourceData) (*serviceendpoint.ServiceEndpoint, *uuid.UUID, error) {
+	serviceEndpoint, projectID := doBaseExpansion(d)
+	serviceEndpoint.Authorization = &serviceendpoint.EndpointAuthorization{
+		Scheme: converter.String("Token"),
+		Parameters: &map[string]string{
+			"apitoken": d.Get("token").(string),
+		},
+	}
+	serviceEndpoint.Type = converter.String("sonarcloud")
+	serviceEndpoint.Url = converter.String("https://sonarcloud.io")
+	return serviceEndpoint, projectID, nil
+}
+
+// Convert AzDO data structure to internal Terraform data structure
+func flattenServiceEndpointSonarCloud(d *schema.ResourceData, serviceEndpoint *serviceendpoint.ServiceEndpoint, projectID *uuid.UUID) {
+	doBaseFlattening(d, serviceEndpoint, projectID)
+
+	tfhelper.HelpFlattenSecret(d, "token")
```

Comment:
`token_hash` can be removed, this help function have length limitation and hash does not do much help. Terraform also prevent to output the sensitive data be default.

Reply:
I have removed the helper function, is there anything else that should be removed?